Alexander Ustinov's pull request

UAlex322 opened the pull request.

UAlex322 committed:
Upload source files

UAlex322 committed:
Upload h-files with class declarations

UAlex322 committed:
Upload files with tests

UAlex322 committed:
Changed test files

UAlex322 committed:
Minor fixes in source files

Changed input/output operators

UAlex322 committed:
Fixed h-files

I had to add binary string constructor for TBitFIeld

ArisovaAnastasia commented:
Есть некоторые замечания. Класс TBitField не собирается. Для использования функции std::max() необходимо подключить библиотеку <algorithm>, где она содержится.

UAlex322 committed:
Fixed tbitfield.cpp

Included <algorithm> to make "std::max" work correctly

ArisovaAnastasia commented on src/tbitfield.cpp:
```diff
@@ -6,81 +6,200 @@
 
 int TBitField::GetMemIndex(const int n) const // индекс Мем для бита n
 {
+	return n/(8*sizeof(TELEM));
```
Данный метод отработает неправильно, если n = -1 или n > Bitlen. В этих случаях необходимо кидать исключение с содержательным комментарием.

ArisovaAnastasia commented on src/tbitfield.cpp:
```diff
@@ -6,81 +6,200 @@
 
 TELEM TBitField::GetMemMask(const int n) const // битовая маска для бита n
 {
+	return static_cast<TELEM>(1 << n % (8*sizeof(TELEM)));
```
Аналогично комментарию выше

ArisovaAnastasia commented on src/tset.cpp:
```diff
@@ -7,82 +7,127 @@
 int TSet::IsMember(const int Elem) const // элемент множества?
 {
-    return 0;
+    return BitField.GetBit(Elem);
```
Желательно выбрасывать исключение с содержательным комментарием, где оно произошло при Elem < 0 и Elem >MaxPower. Иначе будет в таких случаях будет исключение из функции GetBit с комментарием, что произошла ошибка в функции GetBit. Функция GetBit вызывается в нескольких местах в классе TSet, из-за этого поиск где же произошла ошибка сильно затрудняется.

ArisovaAnastasia commented on src/tset.cpp:
```diff
@@ -7,82 +7,127 @@
 
 void TSet::InsElem(const int Elem) // включение элемента множества
 {
+    BitField.SetBit(Elem);
```
Аналогично комментарию выше

ArisovaAnastasia commented on src/tset.cpp:
```diff
@@ -7,82 +7,127 @@
 
 void TSet::DelElem(const int Elem) // исключение элемента множества
 {
+    BitField.ClrBit(Elem);
```
Аналогично комментарию выше

ArisovaAnastasia reviewed (commented):
Есть некоторые замечания, после их исправления лабораторная работа будет принята.

UAlex322 committed:
Fixed exception throwing in some methods

Насчёт методов TBitField - GetMemIndex и GetMemMask - скажу, что они приватные, и извне их нельзя вызвать. Используются они только методами SetBit, GetBit, ClrBit, которые делают проверку на корректность индекса, потому в приватные методы некорректный индекс попасть не может.
В методы TSet - IsMember, InsElem и DelElem - добавил проверку индекса и выбрасывание исключения, если он некорректен.

UAlex322 committed:
Add files via upload

UAlex322 committed:
Fixed TBitField.cpp

Поправил названия методов, выбрасываемые при исключениях (местами было неверно сначала) и исправил GetMemMask (сначала делался static_cast к TELEM результата сдвига единицы типа int, теперь делается сдвиг скастованной к TELEM единицы - в случае с 64-битными целыми метод теперь сработает корректно).

ArisovaAnastasia commented:
Лабораторная работа сдана.
